src/compiler/checker.ts

```diff
@@ -14773,22 +14773,29 @@ namespace ts {
                     // We infer from types that are not naked type variables first so that inferences we
                     // make from nested naked type variables and given slightly higher priority by virtue
                     // of being first in the candidates array.
+                    let typeVariableCount = 0;
                     for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (!getInferenceInfoForType(t)) {
+                        if (getInferenceInfoForType(t)) {
+                            typeVariableCount++;
+                        }
+                        else {
                             inferFromTypes(source, t);
                         }
                     }
                     // Inferences directly to naked type variables are given lower priority as they are
                     // less specific. For example, when inferring from Promise<string> to T | Promise<T>,
-                    // we want to infer string for T, not Promise<string> | string.
-                    const savePriority = priority;
-                    priority |= InferencePriority.NakedTypeVariable;
-                    for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (getInferenceInfoForType(t)) {
-                            inferFromTypes(source, t);
+                    // we want to infer string for T, not Promise<string> | string. For intersection types
+                    // we only infer to single naked type variables.
+                    if (target.flags & TypeFlags.Union ? typeVariableCount !== 0 : typeVariableCount === 1) {

```

Comment:
The old comment on this behavior was
```ts
                    // Next, if target containings a single naked type variable, make a secondary inference to that type
                    // variable. This gives meaningful results for union types in co-variant positions and intersection
                    // types in contra-variant positions (such as callback parameters).
```
should we perhaps be checking variance here, rather than (or perhaps alongside) count?

Reply:
I don't think the `contravariant` flag is relevant here (or maybe I'm just not seeing how).

Reply:
The original in-code comment on this behavior said that it was inferences to intersections in contravariant positions (see above) that gave good results - it's possible that we could capture that sentiment by inferring to only one intersection parameter in covariant positions, but all of them in contravariant positions (and vice-versa for unions, I imagine?)

I dunno really; this is just speculation based on the text of a comment we removed, and assuming that there was insight in that comment which our tests didn't quite capture - without concrete examples, though, I'm not really sure.